core/engine/src/bytecompiler/register.rs

```diff
@@ -0,0 +1,131 @@
+use std::mem::forget;
+
+bitflags::bitflags! {
+    #[derive(Debug, Default, Clone, Copy)]
+    struct RegisterFlags: u8 {
+        /// Whether the register is still in use (not deallocated).
+        const USED       = 0b0000_0001;
+
+        /// Is the register presistent (not deallocatable).
+        const PERSISTENT = 0b0000_0010;
+    }
+}
+
+impl RegisterFlags {
+    fn is_used(self) -> bool {
+        self.contains(Self::USED)
+    }
+    fn is_persistent(self) -> bool {
+        self.contains(Self::PERSISTENT)
+    }
+}
+
+/// An entry in the [`RegisterAllocator`].
+#[derive(Debug, Default)]
+pub(crate) struct RegisterEntry {
+    flags: RegisterFlags,
+}
+
+/// Represent a VM register.
+///
+/// This is intented to be passed by reference or to be moved, dropping this is a bug,
+/// it should only be dropped though the [`RegisterAllocator::dealloc()`] method.
+/// This doesn't apply to persistent registers.
+///
+/// A [`Register`] is index into the register allocator,
+/// as well as an index into the registers on the stack using the register pointer (`rp`).
+#[derive(Debug)]
+pub(crate) struct Register {
+    index: u32,
+    flags: RegisterFlags,
+}
+
+impl Register {
+    /// The index of the [`Register`].
+    pub(crate) fn index(&self) -> u32 {
+        self.index
+    }
+}
+
+impl Drop for Register {
+    fn drop(&mut self) {
+        if self.flags.is_persistent() {
+            return;
+        }
+
+        unreachable!("forgot to deallocate a register! Or a panic happend which caused Reg's drop to be called after the panic!")
+    }
+}
```

Comment:
Seeing this, I wonder if a better approach would be to use drop guards to automatically deallocate the registers, instead of having to mentally juggle when a register is still allocated or not.
You would have to add some lifetimes though, so that the guards are always dropped when you call `finish`.